File: core/src/main/scala/tripaint/coords/GridCoords.scala
package tripaint.coords

case class GridCoords(value: Int) extends AnyVal {
  inline def x: Int = value >> 16
  inline def y: Int = value << 16 >> 16

  def center: (Double, Double) = {
    val vertices: Seq[(Double, Double)] =
      val xDiv2 = (x.toDouble / 2).floor
      val pts =
        if x % 2 == 0 then
          Seq(
            (xDiv2, y),
            (xDiv2 + 1, y),
            (xDiv2, y + 1)
          )
        else
          Seq(
            (xDiv2 + 1, y + 1),
            (xDiv2, y + 1),
            (xDiv2 + 1, y)
          )
      for (xx, yy) <- pts yield (xx + yy * 0.5, -yy * Math.sqrt(3) / 2)

    val centerX: Double = vertices.map(_._1).sum / 3
    val centerY: Double = vertices.map(_._2).sum / 3

    (centerX, centerY)
  }
}

object GridCoords {
  inline def apply(x: Int, y: Int): GridCoords = {
    new GridCoords(x << 16 | (y & 0xffff))
  }
}

File: core/src/main/kotlin/tripaint/coords/GridCoords.kt
package tripaint.coords

import kotlin.math.floor
import kotlin.math.sqrt

data class GridCoords(val value: Int) {
    val x: Int
        get() = value shr 16
    val y: Int
        get() = (value shl 16) shr 16

    val center: Pair<Double, Double>
        get() {
            val xDiv2 = floor(x.toDouble() / 2)
            val pts = if (x % 2 == 0) listOf(
                            Pair(xDiv2, y),
                Pair(xDiv2 + 1, y),
                Pair(xDiv2, y+1)
            )
            else listOf(
                Pair(xDiv2 + 1, y + 1),
                Pair(xDiv2, y+1),
                Pair(xDiv2 + 1, y),
            )
            val vertices: List<Pair<Double, Double>> = pts.map {
                val (xx, yy) = it
                Pair(xx+yy * 0.5, -yy * sqrt(3.0) / 2)
            }

            val centerX: Double = vertices.sumOf { it.first } / 3
            val centerY: Double = vertices.sumOf { it.second } / 3

            return Pair(centerX, centerY)
        }

    companion object {
        fun from(x: Int, y: Int): GridCoords =
            GridCoords((x shl 16) or (y and 0xffff))
    }
}

